### PR TITLE
Unblock JobSchedulerService on shutdown

### LINKED ISSUE #8742: Server doesn't shut down cleanly after service failure during startup
## Expected Behavior
If starting up the `JerseyService` fails, e.g. due to a misconfigured port, the server should initiate a shutdown and the process should terminate.

## Current Behavior
If starting up a the `JerseyService` fails, the server initiates a shutdown. However, it then hangs and the process has to be killed manually.

## Possible Solution

In the stack trace below, the only service mentioned to be not `TERMINATED` is the JobSchedulerService. Looking at a thread dump of the process reveals that we are getting stuck waiting for the server to reach the `RUNNING` state [here](https://github.com/Graylog2/graylog2-server/blob/9a47813f0b9bbe3bbacf545dba4e29ffa8982080/graylog2-server/src/main/java/org/graylog/scheduler/JobSchedulerService.java#L78).
So as the server never reaches that state, we will have to find a way to make the job scheduler shut down nevertheless.

## Steps to Reproduce (for bugs)

1. Change you `graylog.conf` by setting your `http_bind_address` to `http_bind_address = 127.0.0.1:1`
2. Start the Graylog Server
3. Watch the logs to see it shut down after failing to start up correctly

## Context
Stacktrace:
<pre>
2020-08-10 14:13:16,782 ERROR: org.graylog2.bootstrap.ServerBootstrap - Unable to shutdown properly on time. {STOPPING=[JobSchedulerService [STOPPING]], TERMINATED=[InputSetupService [TERMINATED], MongoDBProcessingStatusRecorderService [TERMINATED], UrlWhitelistService [TERMINATED], StreamCacheService [TERMINATED], OutputSetupService [TERMINATED], LookupTableService [TERMINATED], PeriodicalsService [TERMINATED], EtagService [TERMINATED], ConfigurationEtagService [TERMINATED], GracefulShutdownService [TERMINATED], JournalReader [TERMINATED], KafkaJournal [TERMINATED], BufferSynchronizerService [TERMINATED]], FAILED=[JerseyService [FAILED]]}
2020-08-10 14:13:16,782 ERROR: org.graylog2.bootstrap.ServerBootstrap - Graylog startup failed. Exiting. Exception was:
java.lang.IllegalStateException: Expected to be healthy after starting. The following services are not running: {FAILED=[JerseyService [FAILED]]}
	at com.google.common.util.concurrent.ServiceManager$ServiceManagerState.checkHealthy(ServiceManager.java:773) ~[guava-28.2-jre.jar:?]
	at com.google.common.util.concurrent.ServiceManager$ServiceManagerState.awaitHealthy(ServiceManager.java:585) ~[guava-28.2-jre.jar:?]
	at com.google.common.util.concurrent.ServiceManager.awaitHealthy(ServiceManager.java:316) ~[guava-28.2-jre.jar:?]
	at org.graylog2.bootstrap.ServerBootstrap.startCommand(ServerBootstrap.java:161) [classes/:?]
	at org.graylog2.bootstrap.CmdLineTool.run(CmdLineTool.java:210) [classes/:?]
	at org.graylog2.bootstrap.Main.main(Main.java:50) [classes/:?]
	Suppressed: com.google.common.util.concurrent.ServiceManager$FailedService: JerseyService [FAILED]
	Caused by: java.net.SocketException: Permission denied
		at sun.nio.ch.Net.bind0(Native Method) ~[?:1.8.0_222]
		at sun.nio.ch.Net.bind(Net.java:433) ~[?:1.8.0_222]
		at sun.nio.ch.Net.bind(Net.java:425) ~[?:1.8.0_222]
		at sun.nio.ch.ServerSocketChannelImpl.bind(ServerSocketChannelImpl.java:223) ~[?:1.8.0_222]
		at sun.nio.ch.ServerSocketAdaptor.bind(ServerSocketAdaptor.java:74) ~[?:1.8.0_222]
		at org.glassfish.grizzly.nio.transport.TCPNIOBindingHandler.bindToChannelAndAddress(TCPNIOBindingHandler.java:107) ~[grizzly-framework-2.4.4.jar:2.4.4]
		at org.glassfish.grizzly.nio.transport.TCPNIOBindingHandler.bind(TCPNIOBindingHandler.java:64) ~[grizzly-framework-2.4.4.jar:2.4.4]
		at org.glassfish.grizzly.nio.transport.TCPNIOTransport.bind(TCPNIOTransport.java:215) ~[grizzly-framework-2.4.4.jar:2.4.4]
		at org.glassfish.grizzly.nio.transport.TCPNIOTransport.bind(TCPNIOTransport.java:195) ~[grizzly-framework-2.4.4.jar:2.4.4]
		at org.glassfish.grizzly.nio.transport.TCPNIOTransport.bind(TCPNIOTransport.java:186) ~[grizzly-framework-2.4.4.jar:2.4.4]
		at org.glassfish.grizzly.http.server.NetworkListener.start(NetworkListener.java:711) ~[grizzly-http-server-2.4.4.jar:2.4.4]
		at org.glassfish.grizzly.http.server.HttpServer.start(HttpServer.java:256) ~[grizzly-http-server-2.4.4.jar:2.4.4]
		at org.graylog2.shared.initializers.JerseyService.startUpApi(JerseyService.java:201) ~[classes/:?]
		at org.graylog2.shared.initializers.JerseyService.startUp(JerseyService.java:147) ~[classes/:?]
		at com.google.common.util.concurrent.AbstractIdleService$DelegateService$1.run(AbstractIdleService.java:62) ~[guava-28.2-jre.jar:?]
		at com.google.common.util.concurrent.Callables$4.run(Callables.java:119) ~[guava-28.2-jre.jar:?]
		at java.lang.Thread.run(Thread.java:748) ~[?:1.8.0_222]
</pre>

## Your Environment

* Graylog Version: `master`
* Java Version:
* Elasticsearch Version:
* MongoDB Version:
* Operating System:
* Browser version:

### PR DESCRIPTION
**Pull request by thll**
## Description

This change adds an interruptible implementation to wait for the server
reaching the RUNNING state. The new method is then used in the
JobSchedulerService. Upon shutdown, the execution thread of the
JobSchedulerService is interrupted so that the service can
properly terminate.

## Motivation and Context

The JobSchedulerService blocks in its run method until the server
reaches the running state. It cannot be unblocked even when interrupting
the execution thread. So if one of the other services fails during
startup, the service manager is unable to shutdown the
JobSchedulerService.

Fixes #8742 

## How Has This Been Tested?

Manually tested with the following steps:

1. Make the `JerseyService` fail on startup by setting the following `graylog.conf`:
```
http_bind_address = 255.255.255.255:9999
```
2. Start the Graylog server. Without the fix, the server will enter the shutdown sequence but the process will never terminate. With the fix, the server will terminate.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Refactoring (non-breaking change)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.

### CODE REVIEW

**Review by bernd — APPROVED** [review]
LGTM :+1: Thanks!

File: graylog2-server/src/main/java/org/graylog/scheduler/JobSchedulerService.java [hunk 1]
```diff
@@ -69,14 +70,21 @@ public JobSchedulerService(JobExecutionEngine.Factory engineFactory,
...
     @Override
     protected void run() throws Exception {
         // Safety measure to make sure everything is started before we start job scheduling.
         LOG.debug("Waiting for server to enter RUNNING status before starting the scheduler loop");
-        serverStatus.awaitRunning(() -> LOG.debug("Server entered RUNNING state, starting scheduler loop"));
-
+        try {
+            serverStatus.awaitRunning();
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
```
Comment (mpfz0r) [inline graylog2-server/src/main/java/org/graylog/scheduler/JobSchedulerService.java:83]: Hmm,  why do we need this?
The thread is already interrupted. Or am I missing something?

File: graylog2-server/src/main/java/org/graylog2/plugin/ServerStatus.java [hunk 2]
```diff
@@ -149,6 +149,11 @@ public void overrideLoadBalancerThrottled() {
         publishLifecycle(Lifecycle.OVERRIDE_LB_THROTTLED);
     }
 
+    /**
+     * Blocks until the server enters the RUNNING state and then executes the given Runnable.
+     * <p>
+     * <b>This method is not interruptible while waiting for the server to enter the RUNNING state.</b>
+     */
     public void awaitRunning(final Runnable runnable) {
```
Comment (mpfz0r) [inline graylog2-server/src/main/java/org/graylog2/plugin/ServerStatus.java:157]: We could still mark this one `@Deprecated`
```suggestion
    @Deprecated
    public void awaitRunning(final Runnable runnable) {
```

**Review by mpfz0r — CHANGES_REQUESTED** [review]
see comments

File: graylog2-server/src/main/java/org/graylog/scheduler/JobSchedulerService.java (hunk 1 above)
Comment (mpfz0r) [reply graylog2-server/src/main/java/org/graylog/scheduler/JobSchedulerService.java:83]: Nevermind, found the answer myself. Weird semantics :/

Comment (thll) [reply graylog2-server/src/main/java/org/graylog/scheduler/JobSchedulerService.java:83]: Are you referring to setting the interrupt flag with `Thread.currentThread().interrupt();`? In this particular case we don't strictly need it but afaik it's good practice to set the flag when the thread was interrupted to let the caller know about it.

File: graylog2-server/src/main/java/org/graylog2/plugin/ServerStatus.java (hunk 2 above)
Comment (thll) [reply graylog2-server/src/main/java/org/graylog2/plugin/ServerStatus.java:159]: yep, why not

**Review by mpfz0r — APPROVED** [review]
:+1:

**Pull request merged**